x-pack/plugins/fleet/public/applications/integrations/sections/epm/screens/home/card_utils.tsx

```diff
@@ -0,0 +1,232 @@
+  id: string;
+  categories: string[];
+  fromIntegrations?: string;
+  isReauthorizationRequired?: boolean;
+  isUnverified?: boolean;
+  isUpdateAvailable?: boolean;
+  showLabels?: boolean;
+  extraLabelsBadges?: React.ReactNode[];
+}
+
+export const mapToCard = ({
+  getAbsolutePath,
+  getHref,
+  item,
+  addBasePath,
+  packageVerificationKeyId,
+  selectedCategory,
+}: {
+  getAbsolutePath: (p: string) => string;
+  getHref: (page: StaticPage | DynamicPage, values?: DynamicPagePathValues) => string;
+  addBasePath: (url: string) => string;
+  item: CustomIntegration | PackageListItem;
+  packageVerificationKeyId?: string;
+  selectedCategory?: string;
+}): IntegrationCardItem => {
+  let uiInternalPathUrl: string;
+
+  let isUnverified = false;
+
+  const version = 'version' in item ? item.version || '' : '';
+
+  let isUpdateAvailable = false;
+  let isReauthorizationRequired = false;
+  if (item.type === 'ui_link') {
+    uiInternalPathUrl = item.id.includes('language_client.')
+      ? addBasePath(item.uiInternalPath)
+      : item.uiExternalLink || getAbsolutePath(item.uiInternalPath);
+  } else {
+    let urlVersion = item.version;
+    if (item?.installationInfo?.version) {
+      urlVersion = item.installationInfo.version || item.version;
+      isUnverified = isPackageUnverified(item, packageVerificationKeyId);
+      isUpdateAvailable = isPackageUpdatable(item);
+
+      isReauthorizationRequired = hasDeferredInstallations(item);
+    }
+
+    const url = getHref('integration_details_overview', {
+      pkgkey: `${item.name}-${urlVersion}`,
+      ...(item.integration ? { integration: item.integration } : {}),
+    });
+
+    uiInternalPathUrl = url;
+  }
+
+  const release: IntegrationCardReleaseLabel = getPackageReleaseLabel(version);
+
+  let extraLabelsBadges: React.ReactNode[] | undefined;
+  if (item.type === 'integration') {
+    extraLabelsBadges = getIntegrationLabels(item);
+  }
+
+  return {
+    id: `${item.type === 'ui_link' ? 'ui_link' : 'epr'}:${item.id}`,
+    description: item.description,
+    icons: !item.icons || !item.icons.length ? [] : item.icons,
+    title: item.title,
+    url: uiInternalPathUrl,
+    fromIntegrations: selectedCategory,
+    integration: 'integration' in item ? item.integration || '' : '',
+    name: 'name' in item ? item.name : item.id,
+    version,
+    release,
+    categories: ((item.categories || []) as string[]).filter((c: string) => !!c),
+    isReauthorizationRequired,
+    isUnverified,
+    isUpdateAvailable,
+    extraLabelsBadges,
+  };
+};
+
+export function getIntegrationLabels(item: PackageListItem): React.ReactNode[] {
+  const extraLabelsBadges: React.ReactNode[] = [];
+
+  if (
+    item?.installationInfo?.latest_install_failed_attempts?.some(
+      (attempt) =>
+        item.installationInfo && semverLt(item.installationInfo.version, attempt.target_version)
+    )
+  ) {
+    const upgradelFailedAttempt = item.installationInfo?.latest_install_failed_attempts?.find(
```

Comment:
```suggestion
    const upgradeFailedAttempt = item.installationInfo?.latest_install_failed_attempts?.find(
```

Reply:
@kpollich not sure what is the difference here

Reply:
upgradel -> upgrade typo 🙂